checkers/listequals.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package checkers
+
+import (
+	"fmt"
+	"reflect"
+
+	gc "gopkg.in/check.v1"
+)
+
+type listEqualsChecker struct {
+	*gc.CheckerInfo
+}
+
+// The ListEquals checker verifies if two lists are equal. If they are not,
+// it will essentially run a "diff" algorithm to provide the developer with
+// an easily understandable summary of the difference between the two lists.
+var ListEquals gc.Checker = &listEqualsChecker{
+	&gc.CheckerInfo{Name: "ListEquals", Params: []string{"obtained", "expected"}},
+}
+
+func (l *listEqualsChecker) Check(params []interface{}, names []string) (result bool, error string) {
+	obtained := params[0]
+	expected := params[1]
+
+	// Do some simple pre-checks. First, that both 'obtained' and 'expected'
+	// are indeed slices.
+	vExp := reflect.ValueOf(expected)
+	if vExp.Kind() != reflect.Slice {
+		return false, fmt.Sprintf("expected value is not a slice")
+	}
+
+	vObt := reflect.ValueOf(obtained)
+	if vObt.Kind() != reflect.Slice {
+		return false, fmt.Sprintf("obtained value is not a slice")
+	}
+
+	// Check that element types are the same
+	expElemType := vExp.Type().Elem()
+	obtElemType := vObt.Type().Elem()
+
+	if expElemType != obtElemType {
+		return false, fmt.Sprintf("element types are not equal")
+	}
+
+	// Check that the element type is comparable.
+	if !expElemType.Comparable() {
+		return false, fmt.Sprintf("element type is not comparable")
+	}
+
+	// The approach here is to find a longest-common subsequence using dynamic
+	// programming, and use this to generate the diff. This algorithm runs in
+	// O(n^2). However, naive list equality is only O(n). Hence, to be more
+	// efficient, we should first check if the lists are equal, and if they are
+	// not, we do the more complicated work to find out exactly *how* they are
+	// different.
+
+	slicesEqual := true
+	// Check length is equal
+	if vObt.Len() == vExp.Len() {
+		// Iterate through and check every element
+		for i := 0; i < vExp.Len(); i++ {
+			a := vObt.Index(i)
+			b := vExp.Index(i)
+			if !a.Equal(b) {
+				slicesEqual = false
+				break
+			}
+		}
+
+		if slicesEqual {
+			return true, ""
+		}
+	}
+
+	// If we're here, the lists are not equal, so run the DP algorithm to
+	// compute the diff.
+	return false, generateDiff(vObt, vExp)
+}
+
+func generateDiff(obtained, expected reflect.Value) string {
+	// lenLCS[m][n] stores the length of the longest common subsequence of
+	// obtained[:m] and expected[:n]
+	lenLCS := make([][]int, obtained.Len()+1)
```

Comment:
I think we can clean up the memory allocations here. My 20 minute grok of this suggest we are blowing i*j*8 bytes for every check we perform when we can get away with (i+j)*8 bytes.

Reply:
The full table is needed at the end to "trace backwards" and calculate the entire diff. Yes it's O(n*m) in space and time, the onus is on the user to not use this checker for gigantic arrays.

There might be slight optimisations to be made here but I don't think anything more efficient than O(n*m).